### ISSUE TITLE
How to properly use React-Redux typings

### ISSUE THREAD

**Issue by knyzorg** [flow-typed/flow-typed#2553]
I can't seem to find a single guide around properly using react-redux typings.

I am thinking I need _something_ like this:

```jsx
import type { MapStateToProps, MapDispatchToProps } from 'react-redux';

type TMapStateToFilter = MapStateToProps<DataExplorerState, { filters: FilterData[] }, {loading: boolean, filters: FilterData[]}>
const mapStateToFilterProps: TMapStateToFilter = (state) => ({ loading: state.loading, filters: state.filters });


const mapDispatchToFilterProps: MapDispatchToProps<Action, {}, *> = (dispatch) => ({
    updateFilters(filters) { 
        dispatch({ type: "UPDATE_FILTERS", payload: filters }) 
    },
    updateData(data) { 
        dispatch({ type: "UPDATE_DATA", payload: data }) 
    },
    updateLoadState(loading) {
        dispatch({ type: "LOAD", payload: loading })
    }
});

export const FilterBoxConnect = connect(
    mapStateToFilterProps, 
    mapDispatchToFilterProps
)(FilterBox)
```

But it seems wrong as `MapStateToProps` and `MapDispatchToProps` are not exported and I needed to edit the type definition.

There is a preferred way to doing this, right?

**Comment by knyzorg** [comment]
Ohh the connect function is supposed to figure it out. I see.

**Comment by AndrewSouthpaw** [comment]
It Just Works™. 😉

**Comment by knyzorg** [comment]
@AndrewSouthpaw Hey so.. it doesn't seem to Just Work™ for me anymore. I got it to go from an error to an "uncovered code" warning.

Is there some minimal example I can look at which features react-redux with 100% code-coverage? I would share my own code but it's not great as I am just starting to learn react and redux.

**Comment by knyzorg** [comment]
There seems to be a Connector type in react-redux 4, but it went away in 5.

**Comment by AndrewSouthpaw** [comment]
```js
type PropsT = {
  myProp: string,
  // redux
  foo: boolean,
  dispatch: DispatchT,
}

class _Foo extends React.Component<PropsT> {
  render() {
    const { myProp, foo } = this.props

    return (
      <div>{myProp} {foo}</div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    foo: state.foo,
  }
}

export const Foo = connect(mapStateToProps)(_Foo)
```

Note that `state` in the mstp comes in as `any` so you can type it to whatever.

**Comment by AndrewSouthpaw** [comment]
You can also look at the tests for the libdef for examples.

**Comment by knyzorg** [comment]
I copy-pasted the test for the connect method into vscode and Flow says each and every instance of `connect(mapStateToProps, mapDispatchToProps)(Component);` is uncovered.

Your example features the same. Re-opening?

**Comment by AndrewSouthpaw** [comment]
Interesting. Can you provide a repro repo?

**Comment by knyzorg** [comment]
Right here bud :point_right: https://github.com/knyzorg/ReactReduxCoverageBug

Run `flow coverage libdeftests.jsx --color`

**Comment by AndrewSouthpaw** [comment]
Ohhh. You're talking about coverage, I didn't notice that before. Not sure, looping in @villesau in case he has anything to weigh in. The type checking should still work correctly though.

**Comment by EugeneZ** [comment]
I think it might be due to this:

https://github.com/flow-typed/flow-typed/issues/1857#issuecomment-371312890

Basically, since the current `react-redux` types import `Dispatch` and `Store` from outside the `declare module` block, they are typed as `any` *globally* from then on. It's a really nasty bug that I just found today, meaning these types are `any` throughout our app. I moved the definition inside and now I'm getting hundreds of errors.

**Comment by villesau** [comment]
Actually there's even a PR about it: https://github.com/flow-typed/flow-typed/pull/2530

**Comment by knyzorg** [comment]
Am I understanding correctly that the imports are not working as intended? Copy-pasting definitions does not seem like a sane fix if it's flow's imports which need fixing.

**Comment by villesau** [comment]
@knyzorg proper fix would be to move imports inside `declare` but flow-typed does not yer support it. There's movement to fix the issue, but it requires pretty big rewrite as it introduces concept of _dependency_ to flow-typed. This means that when testing flow libdefs, the CI should be able to figure out dependencies and copy them. There are multiple different versions of libdefs (npm version might vary, and flow version might vary) which makes it pretty tricky task to fix properly. Lack of dependencies is currently (imo) top priority project for flow-typed at the moment.

**Comment by callumlocke** [comment]
AFAICT, the react-redux def works fine for stateless functional components, but not for component classes that extend `Component` or `PureComponent`. For v4 anyway.

**Comment by villesau** [comment]
@knyzorg and others, this issue should now be solved! :) see https://github.com/flow-typed/flow-typed/pull/2530

**Comment by LoganBarnett** [comment]
@avilapedro that's 50% better! ;)

I pulled down the project and confirmed the same behavior. I broke apart the expressions and found that `connect` itself is typed, but the result it returns is not known by Flow. I'm baffled by this since all the libdef variants always return something concrete. Maybe the existential type is getting coerced to `any` somehow?

I pulled in the libdef we just merged into my repo and I'm back to having uncovered `connect` results as well. I'm investigating where things went wrong since we'd gotten it up to 100% earlier. It might be related to (but not necessarily the fault of) the `OmitDispatch` that was added just before the merge.

**Comment by LoganBarnett** [comment]
I've narrowed it down to the `$Shape<ST>` thing causing issues. Removing it works for my local repository but it fails the `flow-typed` tests. My local repo uses stateless components exclusively. I know @callumlocke pointed out something similar but I don't yet understand how it relates or how to tackle this issue specifically. I've tried a few things but no joy so far.

@maximeg the test I'm referring to (`testHoistConnectedComponent`) looks like it was added by you. Any thoughts on how we might support both types of components?

**Comment by villesau** [comment]
I'm gonna (re)reopen the issue as there's still some issues and spectacular detective work happening here :)

**Comment by chapati23** [comment]
I'm also very confused. An "official" example would be really helpful.
Just updated from react-redux@4 to react-redux@5 typings and getting loads of errors because `MapStateToProps` and `MapDispatchToProps` are no longer exported by the flow-typed module.

Until now, we did this:

```js
import type { MapStateToProps, MapDispatchToProps } from 'react-redux'
const mapStateToProps: MapStateToProps<StoreStateT, OwnPropsT, StatePropsT> = (
  state: StoreStateT
) => ({ valueFromStore: state.valueFromStore })
```

This no longer works. What's the correct way in the latest version?

**Comment by villesau** [comment]
@chapati23 see this PR: https://github.com/flow-typed/flow-typed/pull/3012 it provides some examples.

**Comment by chapati23** [comment]
Thanks @villesau, that does help a bit! Had never heard of the placeholder `_` type, seems to be documented nowhere :-/

Think i’ll recommend my team to stay on 0.84 a few more months until the dust settles a bit. Flow has definitely been the most painful thing for us to keep up to date in 2018:
* many breaking changes
* no semver
* no explanations for breaking changes (deprecated existential type is still in the docs, placeholder type isn’t, etc..)
* flow-typed packages have a hard time to keep up with the latest flow version as well..

Sorry, bit of a rant here, I know it’s an open source effort but I guess we’ve been spoiled too much by react and expect the same for all other fb-driven projects as well now..

**Comment by villesau** [comment]
@chapati23 flow-typed is not maintained by Facebook, so it's 110% community effort :) Everyone, you and your team included, are more than welcome to contribute library definitions to the project, or even improve the flow-typed tool it self!

What comes to Flow, I hope this clarify a bit: https://github.com/facebook/flow/issues/6833#issuecomment-418608253

Personally I've tried to keep our flow code base (thousands of files typed with Flow) up to date with most recent Flow version. In overall libdefs for most recent versions are usually the best ones available, and updating one version at once has smaller amount of breaking changes so it's easier to update and really put some thought wether the new issues pointed by Flow are valid and how to fix them.

Required annotations (0.85.0) was such a big change that we are still in 0.84.0, but I made an attempt to update to latest without great success yet. Our eslint does not support the latest flow syntax yet, and I believe this will fix the issue: https://github.com/babel/babel-eslint/pull/711 . After babel-eslint is updated, it should be possible to update the Flow. My plan is to suppress new react-redux related errors and let the teams fix them as they go.
